corehq/apps/sms/api.py

```diff
@@ -222,38 +234,44 @@ def process_sms_registration(msg):
     
     To have a mobile worker register itself, do the following:
 
-        NOTE: This is not yet implemented and may change slightly.
-
         1) Select "Enable Mobile Worker Registration via SMS" in project settings.
 
-        2) Text in "join <domain> worker", where <domain> is the domain to join. If the
-        sending number does not exist in the system, a PendingCommCareUser object will be
-        created, tied to that number.
+        2) Text in "join <domain> worker <username>", where <domain> is the domain to join and <username> is the
+        requested username.  If the username doesn't exist it will be created, otherwise the registration will error.
+        If the username argument is not specified, the username will be the mobile number
+
         The "join" and "worker" keywords can be any keyword in REGISTRATION_KEYWORDS and
         REGISTRATION_MOBILE_WORKER_KEYWORDS, respectively. This is meant to support multiple 
         translations.
-
-        3) A domain admin will have to approve the addition of the mobile worker before
-        a CommCareUser can actually be created.
     """
     registration_processed = False
     text_words = msg.text.upper().split()
     keyword1 = text_words[0] if len(text_words) > 0 else ""
     keyword2 = text_words[1].lower() if len(text_words) > 1 else ""
     keyword3 = text_words[2] if len(text_words) > 2 else ""
+    keyword4 = text_words[3] if len(text_words) > 3 else ""
+    cleaned_phone_number = strip_plus(msg.phone_number)
     if keyword1 in REGISTRATION_KEYWORDS and keyword2 != "":
         domain = Domain.get_by_name(keyword2, strict=True)
         if domain is not None:
             if keyword3 in REGISTRATION_MOBILE_WORKER_KEYWORDS and domain.sms_mobile_worker_registration_enabled:
-                #TODO: Register a PendingMobileWorker object that must be approved by a domain admin
-                pass
+                if keyword4 != '':
+                    username = keyword4
+                else:
+                    username = cleaned_phone_number
```

Comment:
This looks good now, thanks.  But one thing I just realized is that we should also do some validation on the username given (the same validation that's done when you create a mobile worker in the ui). 

Also, we're going to have to add the "@[domain.sitename]" to the mobile worker's username.